Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -385,314 +372,62 @@ final class ConnectionPool {
-                case .none:
-                    break
-                }
-                return self.eventLoop.makeFailedFuture(error)
-            }
-        }
-
-        /// Adds a callback on connection close that asks the `state` what to do about this
-        ///
-        /// The callback informs the state about the event, and the state returns a
-        /// `ClosedConnectionRemoveAction` which instructs it about what it should do.
-        private func configureCloseCallback(of connection: Connection) {
-            connection.channel.closeFuture.whenComplete { result in
-                let action: HTTP1ConnectionProvider.State.ClosedConnectionRemoveAction? = self.parentPool.connectionProvidersLock.withLock {
-                    self.stateLock.withLock {
-                        guard connection.mustRunDefaultCloseCallback else {
-                            return nil
-                        }
-                        switch result {
-                        case .success:
-                            return self.state.removeClosedConnection(connection)
-
-                        case .failure(let error):
-                            preconditionFailure("Connection close future failed with error: \(error)")
-                        }
-                    }
-                }
-
-                if let action = action {
-                    switch action {
-                    case .makeConnectionAndComplete(let el, let promise):
-                        self.makeConnection(on: el).cascade(to: promise)
-                    case .none:
-                        break
-                    }
-                }
 
-                connection.closePromise.succeed(())
+                // there is no connection here anymore, we need to bootstrap next waiter
+                self.openedConnectionsCount -= 1
+                self.processNextWaiter()
+                return self.eventLoop.makeFailedFuture(error)
             }
         }
 
-        /// Removes and fails all `waiters`, remove existing `availableConnections` and sets `state.activity` to `.closing`
-        func prepareForClose() -> EventLoopFuture<Void> {
-            let (waitersFutures, closeFutures) = self.stateLock.withLock { () -> ([EventLoopFuture<Connection>], [EventLoopFuture<Void>]) in
-                // Fail waiters
-                let waitersCopy = self.state.waiters
-                self.state.waiters.removeAll()
-                let waitersPromises = waitersCopy.map { $0.promise }
-                let waitersFutures = waitersPromises.map { $0.futureResult }
-                waitersPromises.forEach { $0.fail(HTTPClientError.cancelled) }
-                let closeFutures = self.state.availableConnections.map { $0.close() }
-                return (waitersFutures, closeFutures)
+        func close() {
+            self.lock.withLockVoid {
+                self.waiters.forEach { $0.promise.fail(HTTPClientError.cancelled) }
```

Comment:
this is definitely incorrect. Here we call out under a lock and that must not be the case (and if the test suite should pass we need to add a test here). If anyone were to re-call into AHC from the future that is failed here, we'd get a dead-lock.

If possible, I'd organise the code in a way that it's clear that we're never doing anything to promises whilst the lock is held. A state machine is probably the right tool to make sure that we can safely (under the lock) be transitioned, then returns the action, and finally performs the action (not under a lock).

Reply:
ah, I did miss that, thanks, I'll think about the state machine here, for now I moved stuff out of the locks